Neighbourhood diagonal

niravshah241 opened the pull request:
Once the domain is divided across different MPI ranks, the rank corresponding to the neighbour of a given subdomain in a given direction needs to be identified. Currently the functionality is implemented to identify neighbours in TOP, RIGHT, BOTTOM and LEFT direction which share a common edge with the given subdomain. This Draft PR aims to add functionality to identify neighbours in TOP_RIGHT, BOTTOM_RIGHT, BOTTOM_LEFT and TOP_LEFT corners.

Specifically, the changes include:

- Addition of [is_diagonal_neighbour](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/Partitioner.cpp#L55)
- Addition of [domain_overlap_diagonal](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/DomainUtils.cpp#L37)
- Addition of [Vertex](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/DomainUtils.hpp#L16) for corner neighbours
- Addition of [halo_start_neighbour](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/Partitioner.cpp#L116)
- Addition of [discover_diagonal_neighbours](https://github.com/nextsimhub/domain_decomp/blob/neighbourhood_diagonal/Partitioner.cpp#L493)

This PR has merged the following PRs:
- #81
- #78 
- #77 

It now contains all the functionality required.

Commit:
Diagonal neighbours

Commit:
Diagonal neighbour identification domain_overlap revised

Commit:
Diagonal neighbour identification domain_overlap revised, Tests to be added

TomMelt commented:
@niravshah241 thanks for drafting this PR. I know that it is draft stage but draft PRs still shouldn't contain CMake build files. Please remove these as discussed last meeting. Please only commit changes for relevant source files.

TomMelt commented on DomainUtils.cpp:
```diff
@@ -33,3 +33,33 @@ int domain_overlap(const Domain d1, const Domain d2, const Edge edge)
     return overlap;
 }
+
+bool domain_overlap_diagonal(const Domain d1, const Domain d2, const Vertex vertex)
```
As discussed last meeting, this function `domain_overlap_diagonal` is not needed for the corner neighbours. It can only ever be a fixed size i.e., 1. Please remove this.

niravshah241 replied:
Removed.

TomMelt commented on Partitioner.cpp:
```diff
@@ -433,3 +489,57 @@ void Partitioner::discover_neighbours()
     }
 }
+
+void Partitioner::discover_diagonal_neighbours(){
```
Could you please rename this function to `discover_corner_neighbours`? This will be clearer.

niravshah241 replied:
Removed this function completely and added logic to `discover_neighbours()`.

TomMelt commented on Partitioner.hpp:
```diff
@@ -234,6 +237,7 @@ class LIB_EXPORT Partitioner {
      * @return starting index of halo for the flattened domain array
      */
     int halo_start(const Domain d1, const Domain d2, const Edge edge);
+    int halo_start_neighbour(const Domain d1, const Domain d2, const Vertex vertex);
```
I would also consider renaming this to `halo_start_corner`. We can chat in our next catch up but I think we might not need this in the end either.

niravshah241 replied:
Renamed it to `halo_corner_start`.

TomMelt commented on spack.yaml:
We should look at this `spack.yaml` in a separate pull request

niravshah241 replied:
I removed this spack.yaml from this PR. I had used spack from NextsimDG. It had necessary changes incorporated (see PR https://github.com/nextsimhub/nextsimdg/pull/908) .

TomMelt commented on spack_trial.yaml:
I dont think this should be uploaded to the PR. It looks the same as the other `spack.yaml` file

niravshah241 replied:
Removed

Commit:
Cmake files removed

Commit:
Cleanup:CMakeFiles removed

Commit:
domain_overlap_diagonal from DomainUtils.cpp removed

Commit:
spack.yaml changed

Commit:
domain_overlap_diagonal removed from DomainUtils.hpp

Commit:
discover_corner_neighbour logic updated, neighbour_diagonal namings updated to corner_neighbour

Commit:
Addition of corner_dir_chars, corner_dir_names, Addition of corner neinghbour logic in discover_neighbours

Commit:
Changes in save_metadata for corner neighbours

Commit:
Fix typo in halo_corner_start declaration

Commit:
save_metadata updated

Commit:
save_metadata updated

Commit:
save_metadata updated

dorchard commented on Partitioner.cpp:
```diff
@@ -52,6 +52,40 @@ bool Partitioner::is_neighbour(
     }
 }
 
+bool Partitioner::is_corner_neighbour(
```
Add some documentation here to explain how this works?

dorchard commented on Partitioner.cpp:
```diff
@@ -52,6 +52,40 @@ bool Partitioner::is_neighbour(
+    const Domain d1, const Domain d2, const Vertex vertex, const bool is_px, const bool is_py)
+{
+    if (vertex == TOP_LEFT) {
+        // is_px and is_py is not updatd ath moment as periodic boundaries are not considered.
```
Typo here.

dorchard commented on Partitioner.cpp:
```diff
@@ -79,6 +113,40 @@ int Partitioner::halo_start(const Domain d1, const Domain d2, const Edge edge)
     return start;
 }
 
+int Partitioner::halo_corner_start(const Domain d1, const Domain d2, const Vertex vertex)
```
Make clear that a pre-condition of using this function is that `is_corner_neighbour` is true.

dorchard commented on Partitioner.cpp:
```diff
@@ -252,6 +349,21 @@ void Partitioner::save_metadata(const std::string& filename) const
+        NC_CHECK(nc_def_var(connectivity_gid, (corner_dir_names[vertex] + "_neighbours").c_str(), NC_INT, 1,
+            &corner_dimid, &num_corner_vid[vertex]));
+        // NC_CHECK(nc_def_var(connectivity_gid, (corner_dir_names[vertex] + "_neighbour_ids").c_str(), NC_INT,
+        //     1, &corner_dimids[vertex], &ids_corner_vid[vertex]));
```
As discussed, cleanup this comment (and others).

dorchard commented on Partitioner.cpp:
```diff
@@ -52,6 +52,40 @@ bool Partitioner::is_neighbour(
+{
+    if (vertex == TOP_LEFT) {
+        // is_px and is_py is not updatd ath moment as periodic boundaries are not considered.
+        // Check if TOP Left neighbour i.e., 
```
Some documentation on what `is_px` and `is_py` are for is needed.
I don't understand why the TOP_LEFT case doesn't look at either `is_py` or `is_px` but all the other cases do... why would something be special here?

Commit:
Test for corner neighbour test_3.cdl added

Commit:
Code cleanup

TomMelt commented on DomainUtils.cpp:
@niravshah241 as a rule, I would generally just remove this change. It's just formatting. Unless it's required by the linter (which I dont think it is).

Commit:
Intergration tests updated to include corner neighbour tests, Periodic neighbour tests in integration still need to be fixed, More tests test_3 and test_4 added for corner neighbours, test_1 and test_2 reference files updated to include details on corner neighbour

niravshah241 commented:
Pending tasks:

- [x] Remove minor formatting related errors
- [ ] ~~Add one more test (say test_5) with overlap with other domain on top and bottom edge~~ To be addressed as separately, if necessary
- [x] Rebasing this branch onto main
- [x] Change n,m to x,y in the tests
- [x] Add a comment to the nested functions saying something like, "this should only be called if is_neighbour is true
- [x] Periodic cases (PR https://github.com/nextsimhub/domain_decomp/pull/77)
- [x] Add [Link](https://nextsim-dg.readthedocs.io/en/latest/halo-exchange.html) of Halo exchange logic
- [ ] ~~Unit test for ```is_corner_neighbour```~~ To be addressed as separately and not in this PR

TomMelt commented on Partitioner.cpp:
```diff
@@ -213,6 +302,14 @@ void Partitioner::save_metadata(const std::string& filename) const
+    for (auto vertex : vertices) {
+        NC_CHECK(nc_def_dim(nc_id, corner_dir_chars[vertex].c_str(), corner_dims[vertex], &corner_dimids[vertex]));
+    }
+
```
@niravshah241 you dont need to define `corner_p`/`corner_dimid` because the number of processes is the same whether we have edges or corners. So just use the existing `dimid` variable instead

TomMelt replied:
FYI I came across this whilst adding periodic neighbours so I have fixed it in my PR #77 .

see e.g., https://github.com/nextsimhub/domain_decomp/blob/bef6fc7e1d781d945129a631450a0fcc94d1a700/Partitioner.cpp#L458-L459

TomMelt committed:
Merge branch 'main' into neighbourhood_diagonal

* update tests to work with main branch
* update Partitioner to use new halo_send/recv approach for edges

TomMelt committed:
wip: add periodic neighbour logic

* need to create arrays to store periodic neighbours
* need to output these arrays in Partitioner::save_metadata
* need to add tests (integration)

TomMelt committed:
chore: fix linting

TomMelt committed:
chore: match naming convention of main branch

TomMelt committed:
chore: use std:array to match rest of code

TomMelt committed:
feat: output periodic corner in save_metadata

TomMelt committed:
test: update integration tests with new output

Commit:
Periodic corner neighbour test case

Commit:
Periodic corner neighbour test case

Commit:
Periodic corner neighbour tests test_4_px test_4_py test_4_px_py added

Commit:
Branch unnecesary files cleanup

Commit:
Halo exchange login link added in in haloBufferPositions and comment regarding calling is_corner_neighbour is added in halo_corner_start

Commit:
Code formatting: addition of whitespaces at the end of files and Halo exchange logic comment rephrasing

Commit:
Unit test for is_corner_neighbour and halo_corner_start added

Commit:
Unit test for is_corner_neighbour and halo_corner_start added

Commit:
Periodic unit tests added in test_halo_corner_start and test_is_corner_neighbour

Commit:
halo_corner_start comment corrected

TomMelt committed:
wip: initial fix for non-periodic neighbours

Commit:
test_halo_corner_start updated

Commit:
Partitioner class updated making haloBufferPositions, halo_corner_start, is_neighbour, is_corner_neighbour public members (which were private members earlier)

TomMelt committed:
feat: fix corner neighbour for periodic BC

TomMelt committed:
test: update test reference data

TomMelt committed:
wip: Nirav see if you can find why its breaking

Commit:
Tests updated and fixes in Partition.cpp

Commit:
DOmain decomp test in works

Commit:
Periodic tests added

TomMelt committed:
doc: correct typo in comment regarding haloBufferPositions

TomMelt committed:
feat: fix haloCornerBufferPositions logic

TomMelt committed:
test: generalize test_haloCornerBufferPositions.cpp

TomMelt committed:
test: update reference data for integration test

Commit:
test_haloBufferPositions updated

Commit:
domain decomp image

Commit:
is_neighbour test fixed, corrections in is_neighbour function

TomMelt committed:
chore: use std::vector instead of list

TomMelt committed:
chore: move edge neighbour check logic to is_neighbour

remove neighbour check logic from domain_overlap

update comments in is_neighbour to reflect the changes

in Partitioner we still check the halo_size is non-zero but this will
always be true, so it is not a problem.

TomMelt committed:
chore: update doc for mod function

TomMelt committed:
chore: update filenames in header

TomMelt committed:
test: remove redundant tests

TomMelt committed:
chore: remove unused vars

TomMelt committed:
chore: tidy buffer tests

TomMelt committed:
chore: rename haloBufferPositions to haloEdgeBufferPositions

TomMelt committed:
chore: rename vertices to corners

TomMelt committed:
chore: rename vars to camelCase to match nextsim

TomMelt committed:
chore: apply formatting changes

Co-authored-by: Joe Wallwork <22053413+joewallwork@users.noreply.github.com>

TomMelt commented:
This PR has merged the following PRs:
- #81
- #78 
- #77 

It now contains all the functionality required.

TomMelt reviewed (approved):
All ready to go now.
